cupy/cusparse.py

```diff
@@ -103,6 +118,123 @@ def csrmv(a, x, y=None, alpha=1, beta=0, transa=False):
     return y
 
 
+def csrmvExIsAligned(a, x, y=None):
+    """Check if the pointers of arguments for csrmvEx are aligned or not
+
+    Args:
+        a (cupy.cusparse.csr_matrix): Matrix A.
+        x (cupy.ndarray): Vector x.
+        y (cupy.ndarray or None): Vector y.
+
+        Check if a, x, y pointers are aligned by 128 bytes as
+        required by csrmvEx.
+
+    Returns:
+        bool: ``True`` if all pointers are aligned.
+              ``False`` if otherwise.
+
+    """
+
+    if a.data.data.ptr % 128 != 0:
+        return False
+    if a.indptr.data.ptr % 128 != 0:
+        return False
+    if a.indices.data.ptr % 128 != 0:
+        return False
+    if x.data.ptr % 128 != 0:
+        return False
+    if y is not None and y.data.ptr % 128 != 0:
+        return False
+    return True
+
+
+def _aligned_constant(a, dtype, alignment=128):
+    a = numpy.array(a, dtype)
+    if a.ctypes.data % alignment == 0:
+        # don't waste time if the pointer is aligned
+        return a
+    b = numpy.empty((alignment+a.itemsize, ), 'b')
+    offset = b.ctypes.data % alignment
+    if offset > 0:
+        idx = alignment - offset
+        b = b[idx:]
+    aa = b.view(dtype)
+    aa[0] = a
+    assert aa.ctypes.data % alignment == 0
+    return aa
+
+
+def csrmvEx(a, x, y=None, alpha=1, beta=0, merge_path=True):
+    """Matrix-vector product for a CSR-matrix and a dense vector.
+
+    .. math::
+
+       y = \\alpha * A x + \\beta y,
+
+    Args:
+        a (cupy.cusparse.csr_matrix): Matrix A.
+        x (cupy.ndarray): Vector x.
+        y (cupy.ndarray or None): Vector y. It must be F-contiguous.
+        alpha (float): Coefficient for x.
+        beta (float): Coefficient for y.
+        merge_path (bool): If ``True``, merge path algoritm is used.
+
+        All pointers must be aligned with 128 bytes.
+
+    Returns:
+        cupy.ndarray: Calculated ``y``.
+
+    """
+    assert y is None or y.flags.f_contiguous
+
+    if a.shape[1] != len(x):
+        raise ValueError('dimension mismatch')
+
+    handle = device.get_cusparse_handle()
+    m, n = a.shape
+
+    a, x, y = _cast_common_type(a, x, y)
+    dtype = a.dtype
+    if y is None:
+        y = cupy.zeros(m, dtype)
+
+    datatype = _dtype_to_DataType(dtype)
+    algmode = cusparse.CUSPARSE_ALG_MERGE_PATH if \
+        merge_path else cusparse.CUSPARSE_ALG_NAIVE
+    transa_flag = cusparse.CUSPARSE_OPERATION_NON_TRANSPOSE
+
+    alpha = _aligned_constant(alpha, dtype).ctypes
+    beta = _aligned_constant(beta, dtype).ctypes
+
```

Comment:
According to cusparse documentation examples, it doesnt seem that alpha and beta need to be aligned to 128 bytes although they are being passed as a pointer.

I will probably be wrong but could you double-check this please.

https://docs.nvidia.com/cuda/cusparse/index.html#unique_2009252207

Reply:
Sure.

Be honest, I worked to align alpha and beta too before I realized that
test_csr.py would send a matrix with nnz=0. I may also be wrong. Both of
them are possible reason of illegal memory access error.

Let’s me do more checking tomorrow.

Reply:
Thanks!

Reply:
You are right. I tested with the following:

```python
def _misaligned_constant(a,dtype,alignment=128,misalign=1):
    a = numpy.array(a, dtype)
    if a.ctypes.data % alignment != 0:
        return a

    b = numpy.empty((alignment+a.itemsize+misalign, ), 'b')
    offset = b.ctypes.data % alignment
    if offset > 0:
        idx = alignment - offset + misalign
    else:
        idx = misalign
    b = b[idx:idx+a.itemsize]
    aa = b.view(dtype)
    aa[0] = a
    assert aa.ctypes.data % alignment == misalign
    return aa
```

and use the following instead:

```python
    alpha = _misaligned_constant(alpha, dtype).ctypes
    beta = _misaligned_constant(beta, dtype).ctypes

```

csrmvEx looks good.